Engine/Component/ComponentParticleSystem.cpp

```diff
@@ -0,0 +1,342 @@
+	billboard->ChangeBillboardType(ComponentBillboard::AlignmentType::CROSSED);
+
+	for (unsigned int i = 0; i < max_particles; ++i)
+	{
+		particles.emplace_back(Particle());
+		particles[i].life = 0.0F;
+		particles[i].particle_scale = 1.0F;
+		particles[i].time_passed = particles[i].life;
+	}
+	
+}
+
+unsigned int ComponentParticleSystem::FirstUnusedParticle()
+{
+	for (unsigned int i =last_used_particle; i < max_particles; ++i)
+	{
+		if (particles[i].life <= 0.0f) 
+		{
+			last_used_particle = i;
+			return i;
+		}
+	}
+	for (unsigned int i = 0; i < last_used_particle; ++i) 
+	{
+		if (particles[i].life <= 0.0f) 
+		{
+			last_used_particle = i;
+			return i;
+		}
+	}
+	last_used_particle = 0;
+	return 0;
+}
+
+void ComponentParticleSystem::RespawnParticle(Particle& particle)
+{
+	particle.position = float3(0.0f, 0.0f, 0.0f);
+	particle.rotation = owner->transform.GetGlobalRotation();
+
+	float scale = (rand() % ((max_size_of_particle - min_size_of_particle) + 1) + min_size_of_particle) / 100.f;
+
+	if (size_random)
+	{
+		particle.particle_scale = scale;
+	}
+	else 
+	{
+		particle.particle_scale = 1.0f;
+	}
+
+	switch (type_of_particle_system)
+	{
+		case SPHERE:
+			particle.velocity = float3::RandomDir(LCG(), velocity_particles) / 1000;
+		break;
+		case BOX:
+		{
+			float random_x = (rand() % ((max_range_random_x - min_range_random_x) + 1) + min_range_random_x) / 100.f;
+
+			float random_z = (rand() % ((max_range_random_z - min_range_random_z) + 1) + min_range_random_z) / 100.f;
+
+			particle.velocity.y = velocity_particles / 1000;
+			if (enabled_random_x)
+			{
+				particle.position.x = random_x;
+			}
+			else
+			{
+				particle.position.x = position_x / 100.0F;
+			}
+			if (enabled_random_z)
+			{
+				particle.position.z = random_z;
+			}
+			else
+			{
+				particle.position.z = position_z / 100.0F;
+			}
+		
+		break;
+		}
+		case CONE:
+			float angle = ((rand() % 100)/100.f) * 2 * math::pi;
+			float radius = inner_radius * sqrt((rand() % 100) / 100.f);
+			float proportion = outer_radius / inner_radius;
+			particle.position.x = radius * math::Cos(angle);
+			particle.position.z = radius * math::Sin(angle);
+			float distance = velocity_particles * particles_life_time;
+			float height = sqrt((distance*distance) - ((radius*proportion) - radius)*((radius*proportion) - radius));
+			float3 final_local_position = float3(particle.position.x*proportion, height, particle.position.z*proportion);
+			particle.velocity = (final_local_position - particle.position);
+			particle.velocity = particle.velocity.ScaledToLength(velocity_particles) / 1000;
+		break;
+
+	}
+	
+	particle.color = { color_particle[0], color_particle[1], color_particle[2], 1.0f };
+	particle.life = particles_life_time*1000;
+	particle.time_passed = particle.life;
+	float4 aux_velocity(particle.velocity, 1.0F);
+	aux_velocity = particle.rotation * aux_velocity;
+	particle.velocity = aux_velocity.xyz();
+
+	if (follow_owner)
+	{
```

Comment:
why is this block empty?

Reply:
you are right i should use there if(!follow_owner)